stdlib/asm/crypto/dsa/rpo_falcon512.masm

```diff
@@ -484,65 +520,89 @@ end
         mem_loadw
-        #=> [PI, PI_{i+512}, C, pi_ptr, ...]
+        #=> [C, PI_{i+512}, PI, pi_ptr, ...]
 
         # 2) Compute the squared norm of (i + 0)-th coefficient of s1
-        movup.8
-        exec.mod_12289
-        movup.5
-        #=> [v, w, u, ...] where u is the i-th coefficient of `pi`, v is the i-th
-        # coefficient of `c` and w is the (512 + i)-th coefficient of `pi` polynomial.
 
+        u32split
+        #=> [c0_hi, c0_lo, c1, c2, c3, PI_{i+512}, PI, pi_ptr, ...]
+        movup.5
+        push.J add
+        #=> [pi512_0, c0_hi, c0_lo, c1, c2, c3, pi512_1, pi512_2, pi512_3, PI, pi_ptr, ...]
+        movup.9
+        #=> [pi0, pi512_0, c_hi, c_lo, c1, c2, c3, pi512_1, pi512_2, pi512_3, pi1, pi2, pi3, pi_ptr, ...]
         exec.diff_mod_q
-        #=> [e, ...]
-
+        #=> [e, c1, c2, c3, pi512_1, pi512_2, pi512_3, pi1, pi2, pi3, pi_ptr, ...]
         exec.norm_sq
         #=> [norm(e)^2, ...]
 
         # Move the result out of the way so that we can process the remaining coefficients
         movdn.10
+        #=> [c1, c2, c3, pi512_1, pi512_2, pi512_3, pi1, pi2, pi3, pi_ptr, e0, ...]
 
         # 3) Compute the squared norm of (i + 1)-th coefficient of s1
-        movup.6
-        exec.mod_12289
+
+        u32split
+        #=> [c1_hi, c1_lo, c2, c3, pi512_1, pi512_2, pi512_3, pi1, pi2, pi3, pi_ptr, e0, ...]
         movup.4
+        #=> [pi512_1, c1_hi, c1_lo, c2, c3, pi512_2, pi512_3, pi2, pi3, pi_ptr, e0, ...]
+        push.J add
+        movup.7
+        #=> [pi1, pi512_1, c1_hi, c1_lo, c2, c3, pi512_2, pi512_3, pi2, pi3, pi_ptr, e0, ...]
         exec.diff_mod_q
         exec.norm_sq
+        #=> [e, c2, c3, pi512_2, pi512_3, pi2, pi3, pi_ptr, e0, ...]
+
         movdn.7
+        #=> [c2, c3, pi512_2, pi512_3, pi2, pi3, pi_ptr, e0, e1, ...]
 
         # 4) Compute the squared norm of (i + 2)-th coefficient of s1
-        movup.4
-        exec.mod_12289
+
+        u32split
+        #=> [c2_hi, c2_lo, c3, pi512_2, pi512_3, pi2, pi3, pi_ptr, e0, e1, ...]
         movup.3
+        push.J add
+        #=> [pi512_2, c2_hi, c2_lo, c3, pi512_3, pi2, pi3, pi_ptr, e0, e1, ...]
+        movup.5
+        #=> [pi2, pi512_2, c2_hi, c2_lo, c3, pi512_3, pi3, pi_ptr, e0, e1, ...]
         exec.diff_mod_q
         exec.norm_sq
+
         movdn.4
+        #=> [c3, pi512_3, pi3, pi_ptr, e, e, e, ...]
 
         # 5) Compute the squared norm of (i + 3)-th coefficient of s1
-        movup.2
-        exec.mod_12289
-        movup.2
+
+        u32split
+        #=> [c3_hi, c3_lo, pi512_3, pi3, pi_ptr, e0, e1, e2, ...]
+        movup.2 push.J add
+        movup.3
+        #=> [pi3, pi512_3, c3_hi, c3_lo, pi_ptr, e0, e1, e2, ...]
         exec.diff_mod_q
+        #=> [e3, pi_ptr, e0, e1, e2, ...]
         exec.norm_sq
+        #=> [e3, pi_ptr, e0, e1, e2, ...]
+
         swap
+        #=> [pi_ptr, e3, e0, e1, e2, ...]
 
         # 6) Increment the pointer
         add.1
```

Comment:
Similar to the previous comment: should this not be `add.4`?

Reply:
Updated